File: agents/agent_service/lib/memory_service.py
import os
import time
import json
import logging
from typing import Dict, Any, List, Optional, Tuple
import httpx
from uuid import uuid4
import asyncio
import redis.asyncio as redis
from dotenv import load_dotenv

# Load environment variables
load_dotenv()

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)

# Get environment variables
REDIS_URL = os.getenv("REDIS_URL")
PINECONE_API_KEY = os.getenv("PINECONE_API_KEY")
PINECONE_ENVIRONMENT = os.getenv("PINECONE_ENVIRONMENT")
GEMINI_API_KEY = os.getenv("GEMINI_API_KEY")

class MemoryService:
    """Service for storing and retrieving agent memory."""
    
    def __init__(self):
        """Initialize the memory service."""
        self.redis_client = None
        self.pinecone_client = None
        self.memory_cache = {}  # In-memory fallback
        
        # Initialize Redis if URL is provided
        if REDIS_URL and not REDIS_URL.startswith("your_"):
            try:
                self.redis_client = redis.from_url(REDIS_URL)
                logger.info("✅ Connected to Redis for memory service")
            except Exception as e:
                logger.error(f"❌ Failed to connect to Redis: {str(e)}")
                logger.info("⚠️ Using in-memory cache for memory service")
        else:
            logger.info("⚠️ Redis URL not provided, using in-memory cache")
        
        # Initialize Pinecone client if API key is provided
        # In a full implementation, we would connect to Pinecone for vector storage
        if PINECONE_API_KEY and not PINECONE_API_KEY.startswith("your_"):
            logger.info("✅ Pinecone API key found for long-term memory")
        else:
            logger.info("⚠️ Pinecone not configured, long-term memory will be limited")
    
    async def store_memory(
        self,
        agent_id: str,
        content: str,
        memory_type: str = "interaction",
        metadata: Optional[Dict[str, Any]] = None,
        importance: float = 0.5,
        user_id: Optional[str] = None,
        expiration: Optional[int] = None
    ) -> str:
        """Store a memory for an agent.
        
        Args:
            agent_id: The ID of the agent.
            content: The content of the memory.
            memory_type: The type of memory (interaction, learning, etc.).
            metadata: Additional metadata about the memory.
            importance: Importance score (0-1).
            user_id: User ID associated with this memory.
            expiration: Optional TTL in seconds.
            
        Returns:
            The memory ID.
        """
        memory_id = str(uuid4())
        timestamp = int(time.time())
        
        memory = {
            "id": memory_id,
            "agent_id": agent_id,
            "content": content,
            "type": memory_type,
            "metadata": metadata or {},
            "importance": importance,
            "created_at": timestamp,
            "user_id": user_id
        }
        
        # Try to store in Redis first
        if self.redis_client:
            try:
                key = f"memory:{agent_id}:{memory_id}"
                await self.redis_client.set(key, json.dumps(memory))
                
                # Set expiration if specified
                if expiration:
                    await self.redis_client.expire(key, expiration)
                
                # Add to memory index for agent
                await self.redis_client.zadd(
                    f"memory_index:{agent_id}",
                    {memory_id: timestamp}
                )
                
                # Add to importance index
                await self.redis_client.zadd(
                    f"memory_importance:{agent_id}",
                    {memory_id: importance}
                )
                
                logger.info(f"✅ Memory {memory_id} stored in Redis for agent {agent_id}")
            except Exception as e:
                logger.error(f"❌ Failed to store memory in Redis: {str(e)}")
                # Fall back to in-memory storage
                self._store_in_memory(agent_id, memory_id, memory)
        else:
            # Store in memory
            self._store_in_memory(agent_id, memory_id, memory)
        
        # In a full implementation, we would also store in Pinecone for vector search
        return memory_id
    
    def _store_in_memory(self, agent_id: str, memory_id: str, memory: Dict[str, Any]):
        """Store memory in the in-memory cache.
        
        Args:
            agent_id: The agent ID.
            memory_id: The memory ID.
            memory: The memory data.
        """
        if agent_id not in self.memory_cache:
            self.memory_cache[agent_id] = {}
        
        self.memory_cache[agent_id][memory_id] = memory
        logger.info(f"✅ Memory {memory_id} stored in-memory for agent {agent_id}")
    
    async def retrieve_recent_memories(
        self, 
        agent_id: str, 
        limit: int = 10
    ) -> List[Dict[str, Any]]:
        """Retrieve the most recent memories for an agent.
        
        Args:
            agent_id: The ID of the agent.
            limit: Maximum number of memories to retrieve.
            
        Returns:
            List of memory objects.
        """
        memories = []
        
        # Try to get from Redis first
        if self.redis_client:
            try:
                # Get memory IDs from the index, sorted by timestamp (newest first)
                memory_ids = await self.redis_client.zrevrange(
                    f"memory_index:{agent_id}", 
                    0, 
                    limit - 1
                )
                
                # Get the actual memories
                for memory_id in memory_ids:
                    memory_id_str = memory_id.decode("utf-8") if isinstance(memory_id, bytes) else memory_id
                    memory_json = await self.redis_client.get(f"memory:{agent_id}:{memory_id_str}")
                    if memory_json:
                        memories.append(json.loads(memory_json))
                
                logger.info(f"✅ Retrieved {len(memories)} recent memories from Redis for agent {agent_id}")
            except Exception as e:
                logger.error(f"❌ Failed to retrieve memories from Redis: {str(e)}")
                # Fall back to in-memory retrieval
                memories = self._retrieve_from_memory(agent_id, sort_by="timestamp", limit=limit)
        else:
            # Get from in-memory cache
            memories = self._retrieve_from_memory(agent_id, sort_by="timestamp", limit=limit)
        
        return memories
    
    async def retrieve_important_memories(
        self, 
        agent_id: str, 
        limit: int = 10
    ) -> List[Dict[str, Any]]:
        """Retrieve the most important memories for an agent.
        
        Args:
            agent_id: The ID of the agent.
            limit: Maximum number of memories to retrieve.
            
        Returns:
            List of memory objects.
        """
        memories = []
        
        # Try to get from Redis first
        if self.redis_client:
            try:
                # Get memory IDs from the importance index, sorted by importance (highest first)
                memory_ids = await self.redis_client.zrevrange(
                    f"memory_importance:{agent_id}", 
                    0, 
                    limit - 1
                )
                
                # Get the actual memories
                for memory_id in memory_ids:
                    memory_id_str = memory_id.decode("utf-8") if isinstance(memory_id, bytes) else memory_id
                    memory_json = await self.redis_client.get(f"memory:{agent_id}:{memory_id_str}")
                    if memory_json:
                        memories.append(json.loads(memory_json))
                
                logger.info(f"✅ Retrieved {len(memories)} important memories from Redis for agent {agent_id}")
            except Exception as e:
                logger.error(f"❌ Failed to retrieve important memories from Redis: {str(e)}")
                # Fall back to in-memory retrieval
                memories = self._retrieve_from_memory(agent_id, sort_by="importance", limit=limit)
        else:
            # Get from in-memory cache
            memories = self._retrieve_from_memory(agent_id, sort_by="importance", limit=limit)
        
        return memories
    
    def _retrieve_from_memory(
        self, 
        agent_id: str, 
        sort_by: str = "timestamp", 
        limit: int = 10
    ) -> List[Dict[str, Any]]:
        """Retrieve memories from the in-memory cache.
        
        Args:
            agent_id: The agent ID.
            sort_by: Field to sort by ('timestamp' or 'importance').
            limit: Maximum number of memories to retrieve.
            
        Returns:
            List of memory objects.
        """
        if agent_id not in self.memory_cache:
            return []
        
        # Get all memories for the agent
        agent_memories = list(self.memory_cache[agent_id].values())
        
        # Sort based on the specified field
        if sort_by == "timestamp":
            agent_memories.sort(key=lambda x: x.get("created_at", 0), reverse=True)
        elif sort_by == "importance":
            agent_memories.sort(key=lambda x: x.get("importance", 0), reverse=True)
        
        # Return limited number of memories
        return agent_memories[:limit]
    
    async def search_memories(
        self, 
        agent_id: str, 
        query: str, 
        limit: int = 5
    ) -> List[Dict[str, Any]]:
        """Search agent memories based on content similarity.
        
        Args:
            agent_id: The ID of the agent.
            query: The search query.
            limit: Maximum number of results.
            
        Returns:
            List of memory objects.
        """
        # In a full implementation, we would use vector similarity search with Pinecone
        # For now, we'll do a simple keyword search
        
        # Convert query to lowercase for case-insensitive matching
        query_lower = query.lower()
        
        # Get all memories for the agent
        if self.redis_client:
            try:
                # Get all memory IDs for the agent
                memory_ids = await self.redis_client.zrange(
                    f"memory_index:{agent_id}", 
                    0, 
                    -1
                )
                
                memories = []
                for memory_id in memory_ids:
                    memory_id_str = memory_id.decode("utf-8") if isinstance(memory_id, bytes) else memory_id
                    memory_json = await self.redis_client.get(f"memory:{agent_id}:{memory_id_str}")
                    if memory_json:
                        memory = json.loads(memory_json)
                        memories.append(memory)
                
                # Filter memories that contain the query in the content
                results = [
                    memory for memory in memories
                    if query_lower in memory.get("content", "").lower()
                ]
                
                # Sort by relevance (simple implementation)
                results.sort(
                    key=lambda x: (
                        query_lower in x.get("content", "").lower(),
                        x.get("importance", 0)
                    ),
                    reverse=True
                )
                
                return results[:limit]
            except Exception as e:
                logger.error(f"❌ Failed to search memories in Redis: {str(e)}")
                # Fall back to in-memory search
                return self._search_in_memory(agent_id, query_lower, limit)
        else:
            # Search in-memory cache
            return self._search_in_memory(agent_id, query_lower, limit)
    
    def _search_in_memory(self, agent_id: str, query: str, limit: int) -> List[Dict[str, Any]]:
        """Search memories in the in-memory cache.
        
        Args:
            agent_id: The agent ID.
            query: The search query (lowercase).
            limit: Maximum number of results.
            
        Returns:
            List of memory objects.
        """
        if agent_id not in self.memory_cache:
            return []
        
        # Get all memories for the agent
        agent_memories = list(self.memory_cache[agent_id].values())
        
        # Filter memories that contain the query in the content
        results = [
            memory for memory in agent_memories
            if query in memory.get("content", "").lower()
        ]
        
        # Sort by relevance (simple implementation)
        results.sort(
            key=lambda x: (
                query in x.get("content", "").lower(),
                x.get("importance", 0)
            ),
            reverse=True
        )
        
        return results[:limit]
    
    async def delete_memory(self, agent_id: str, memory_id: str) -> bool:
        """Delete a memory.
        
        Args:
            agent_id: The ID of the agent.
            memory_id: The ID of the memory to delete.
            
        Returns:
            True if successful, False otherwise.
        """
        # Try to delete from Redis first
        if self.redis_client:
            try:
                # Remove from memory storage
                await self.redis_client.delete(f"memory:{agent_id}:{memory_id}")
                
                # Remove from indices
                await self.redis_client.zrem(f"memory_index:{agent_id}", memory_id)
                await self.redis_client.zrem(f"memory_importance:{agent_id}", memory_id)
                
                logger.info(f"✅ Memory {memory_id} deleted from Redis for agent {agent_id}")
                
                # Also remove from in-memory cache if it exists there
                if agent_id in self.memory_cache and memory_id in self.memory_cache[agent_id]:
                    del self.memory_cache[agent_id][memory_id]
                
                return True
            except Exception as e:
                logger.error(f"❌ Failed to delete memory from Redis: {str(e)}")
                # Fall back to in-memory deletion
                return self._delete_from_memory(agent_id, memory_id)
        else:
            # Delete from in-memory cache
            return self._delete_from_memory(agent_id, memory_id)
    
    def _delete_from_memory(self, agent_id: str, memory_id: str) -> bool:
        """Delete memory from the in-memory cache.
        
        Args:
            agent_id: The agent ID.
            memory_id: The memory ID.
            
        Returns:
            True if successful, False otherwise.
        """
        if agent_id not in self.memory_cache:
            return False
        
        if memory_id in self.memory_cache[agent_id]:
            del self.memory_cache[agent_id][memory_id]
            logger.info(f"✅ Memory {memory_id} deleted from in-memory cache for agent {agent_id}")
            return True
        
        return False
    
    async def clear_agent_memories(self, agent_id: str) -> bool:
        """Clear all memories for an agent.
        
        Args:
            agent_id: The ID of the agent.
            
        Returns:
            True if successful, False otherwise.
        """
        # Try to clear from Redis first
        if self.redis_client:
            try:
                # Get all memory IDs for the agent
                memory_ids = await self.redis_client.zrange(
                    f"memory_index:{agent_id}", 
                    0, 
                    -1
                )
                
                # Delete all memories
                for memory_id in memory_ids:
                    memory_id_str = memory_id.decode("utf-8") if isinstance(memory_id, bytes) else memory_id
                    await self.redis_client.delete(f"memory:{agent_id}:{memory_id_str}")
                
                # Delete indices
                await self.redis_client.delete(f"memory_index:{agent_id}")
                await self.redis_client.delete(f"memory_importance:{agent_id}")
                
                logger.info(f"✅ All memories cleared for agent {agent_id}")
                
                # Also clear from in-memory cache
                if agent_id in self.memory_cache:
                    del self.memory_cache[agent_id]
                
                return True
            except Exception as e:
                logger.error(f"❌ Failed to clear memories from Redis: {str(e)}")
                # Fall back to in-memory clearing
                return self._clear_memory_cache(agent_id)
        else:
            # Clear from in-memory cache
            return self._clear_memory_cache(agent_id)
    
    def _clear_memory_cache(self, agent_id: str) -> bool:
        """Clear agent memories from the in-memory cache.
        
        Args:
            agent_id: The agent ID.
            
        Returns:
            True if successful, False otherwise.
        """
        if agent_id in self.memory_cache:
            del self.memory_cache[agent_id]
            logger.info(f"✅ All memories cleared from in-memory cache for agent {agent_id}")
            return True
        
        return False
    
    async def get_memory(self, agent_id: str, memory_id: str) -> Optional[Dict[str, Any]]:
        """Get a specific memory by ID.
        
        Args:
            agent_id: The ID of the agent.
            memory_id: The ID of the memory.
            
        Returns:
            The memory object or None if not found.
        """
        # Try to get from Redis first
        if self.redis_client:
            try:
                memory_json = await self.redis_client.get(f"memory:{agent_id}:{memory_id}")
                if memory_json:
                    return json.loads(memory_json)
                
                logger.warning(f"⚠️ Memory {memory_id} not found in Redis for agent {agent_id}")
            except Exception as e:
                logger.error(f"❌ Failed to get memory from Redis: {str(e)}")
        
        # Fall back to in-memory cache
        if agent_id in self.memory_cache and memory_id in self.memory_cache[agent_id]:
            return self.memory_cache[agent_id][memory_id]
        
        return None
    
    async def update_memory_importance(
        self, 
        agent_id: str, 
        memory_id: str, 
        importance: float
    ) -> bool:
        """Update the importance score of a memory.
        
        Args:
            agent_id: The ID of the agent.
            memory_id: The ID of the memory.
            importance: New importance score (0-1).
            
        Returns:
            True if successful, False otherwise.
        """
        # Try to update in Redis first
        if self.redis_client:
            try:
                # Get the memory
                memory_json = await self.redis_client.get(f"memory:{agent_id}:{memory_id}")
                if not memory_json:
                    logger.warning(f"⚠️ Memory {memory_id} not found in Redis for agent {agent_id}")
                    return False
                
                memory = json.loads(memory_json)
                memory["importance"] = importance
                
                # Update the memory
                await self.redis_client.set(f"memory:{agent_id}:{memory_id}", json.dumps(memory))
                
                # Update the importance index
                await self.redis_client.zadd(
                    f"memory_importance:{agent_id}",
                    {memory_id: importance}
                )
                
                logger.info(f"✅ Importance updated to {importance} for memory {memory_id}")
                
                # Also update in in-memory cache if it exists there
                if agent_id in self.memory_cache and memory_id in self.memory_cache[agent_id]:
                    self.memory_cache[agent_id][memory_id]["importance"] = importance
                
                return True
            except Exception as e:
                logger.error(f"❌ Failed to update memory importance in Redis: {str(e)}")
                # Fall back to in-memory update
                return self._update_importance_in_memory(agent_id, memory_id, importance)
        else:
            # Update in in-memory cache
            return self._update_importance_in_memory(agent_id, memory_id, importance)
    
    def _update_importance_in_memory(self, agent_id: str, memory_id: str, importance: float) -> bool:
        """Update memory importance in the in-memory cache.
        
        Args:
            agent_id: The agent ID.
            memory_id: The memory ID.
            importance: New importance score.
            
        Returns:
            True if successful, False otherwise.
        """
        if agent_id not in self.memory_cache or memory_id not in self.memory_cache[agent_id]:
            return False
        
        self.memory_cache[agent_id][memory_id]["importance"] = importance
        logger.info(f"✅ Importance updated to {importance} for memory {memory_id} in in-memory cache")
        return True
    
    async def close(self):
        """Close connections to external services."""
        if self.redis_client:
            await self.redis_client.close()
            logger.info("✅ Redis client closed")


# Create a singleton instance for the service
_memory_service = None

def get_memory_service() -> MemoryService:
    """Get the singleton MemoryService instance.
    
    Returns:
        MemoryService instance.
    """
    global _memory_service
    if _memory_service is None:
        _memory_service = MemoryService()
    return _memory_service

File: orchestrator/src/services/memoryService.ts
import { createClient as createRedisClient, RedisClientType } from 'redis';
import { v4 as uuidv4 } from 'uuid';
import dotenv from 'dotenv';

// Load environment variables
dotenv.config();

// Redis URL from environment
const REDIS_URL = process.env.REDIS_URL || 'redis://localhost:6379';

// Interface for Memory
interface Memory {
  id: string;
  agent_id: string;
  content: string;
  type: string;
  metadata?: Record<string, any>;
  importance: number;
  created_at: number;
  user_id?: string;
  expires_at?: number;
}

// Interface for Memory Search Result
interface MemorySearchResult {
  memory: Memory;
  score: number;
}

class MemoryService {
  private redisClient: RedisClientType | null = null;
  private inMemoryCache: Record<string, Record<string, Memory>> = {};
  private isRedisConnected: boolean = false;

  constructor() {
    this.initRedis();
  }

  private async initRedis() {
    // Skip Redis initialization if URL is a placeholder or undefined
    if (!REDIS_URL || REDIS_URL.includes('your_redis') || REDIS_URL === 'redis://localhost:6379') {
      console.log('⚠️ Redis URL not provided or is placeholder. Using in-memory storage for agent memories.');
      return;
    }

    try {
      this.redisClient = createRedisClient({ url: REDIS_URL });
      await this.redisClient.connect();
      
      this.redisClient.on('error', (err) => {
        console.error('❌ Redis client error:', err);
        this.isRedisConnected = false;
      });
      
      this.redisClient.on('ready', () => {
        console.log('✅ Redis client connected');
        this.isRedisConnected = true;
      });
      
      console.log('✅ Redis memory service initialized');
      this.isRedisConnected = true;
    } catch (error) {
      console.error('❌ Failed to connect to Redis:', error);
      console.log('⚠️ Using in-memory storage for agent memories');
      this.isRedisConnected = false;
    }
  }

  /**
   * Store a memory for an agent
   */
  public async storeMemory(
    agent_id: string,
    content: string,
    type: string = 'interaction',
    metadata: Record<string, any> = {},
    importance: number = 0.5,
    user_id?: string,
    expiration?: number
  ): Promise<string> {
    const memory_id = uuidv4();
    const timestamp = Date.now();
    
    const memory: Memory = {
      id: memory_id,
      agent_id,
      content,
      type,
      metadata,
      importance,
      created_at: timestamp,
      user_id,
      expires_at: expiration ? timestamp + (expiration * 1000) : undefined
    };
    
    // Try to store in Redis if connected
    if (this.isRedisConnected && this.redisClient) {
      try {
        const key = `memory:${agent_id}:${memory_id}`;
        await this.redisClient.set(key, JSON.stringify(memory));
        
        // Set expiration if specified
        if (expiration) {
          await this.redisClient.expire(key, expiration);
        }
        
        // Add to memory index for agent (sorted by timestamp)
        await this.redisClient.zAdd(`memory_index:${agent_id}`, {
          score: timestamp,
          value: memory_id
        });
        
        // Add to importance index
        await this.redisClient.zAdd(`memory_importance:${agent_id}`, {
          score: importance,
          value: memory_id
        });
        
        console.log(`✅ Memory ${memory_id} stored in Redis for agent ${agent_id}`);
      } catch (error) {
        console.error(`❌ Failed to store memory in Redis: ${error}`);
        // Fall back to in-memory storage
        this.storeInMemory(agent_id, memory_id, memory);
      }
    } else {
      // Store in memory
      this.storeInMemory(agent_id, memory_id, memory);
    }
    
    return memory_id;
  }

  /**
   * Store a memory in the in-memory cache
   */
  private storeInMemory(agent_id: string, memory_id: string, memory: Memory): void {
    if (!this.inMemoryCache[agent_id]) {
      this.inMemoryCache[agent_id] = {};
    }
    
    this.inMemoryCache[agent_id][memory_id] = memory;
    console.log(`✅ Memory ${memory_id} stored in-memory for agent ${agent_id}`);
  }

  /**
   * Retrieve recent memories for an agent
   */
  public async retrieveRecentMemories(
    agent_id: string,
    limit: number = 10
  ): Promise<Memory[]> {
    // Try to get from Redis if connected
    if (this.isRedisConnected && this.redisClient) {
      try {
        // Get memory IDs from the index, sorted by timestamp (newest first)
        const memory_ids = await this.redisClient.zRange(
          `memory_index:${agent_id}`,
          0,
          limit - 1,
          { REV: true }
        );
        
        if (!memory_ids || memory_ids.length === 0) {
          return [];
        }
        
        // Get the actual memories
        const memories: Memory[] = [];
        
        for (const memory_id of memory_ids) {
          const memory_json = await this.redisClient.get(`memory:${agent_id}:${memory_id}`);
          
          if (memory_json) {
            memories.push(JSON.parse(memory_json));
          }
        }
        
        console.log(`✅ Retrieved ${memories.length} recent memories from Redis for agent ${agent_id}`);
        return memories;
      } catch (error) {
        console.error(`❌ Failed to retrieve memories from Redis: ${error}`);
        // Fall back to in-memory retrieval
        return this.retrieveFromMemory(agent_id, 'timestamp', limit);
      }
    } else {
      // Get from in-memory cache
      return this.retrieveFromMemory(agent_id, 'timestamp', limit);
    }
  }

  /**
   * Retrieve important memories for an agent
   */
  public async retrieveImportantMemories(
    agent_id: string,
    limit: number = 10
  ): Promise<Memory[]> {
    // Try to get from Redis if connected
    if (this.isRedisConnected && this.redisClient) {
      try {
        // Get memory IDs from the importance index, sorted by importance (highest first)
        const memory_ids = await this.redisClient.zRange(
          `memory_importance:${agent_id}`,
          0,
          limit - 1,
          { REV: true }
        );
        
        if (!memory_ids || memory_ids.length === 0) {
          return [];
        }
        
        // Get the actual memories
        const memories: Memory[] = [];
        
        for (const memory_id of memory_ids) {
          const memory_json = await this.redisClient.get(`memory:${agent_id}:${memory_id}`);
          
          if (memory_json) {
            memories.push(JSON.parse(memory_json));
          }
        }
        
        console.log(`✅ Retrieved ${memories.length} important memories from Redis for agent ${agent_id}`);
        return memories;
      } catch (error) {
        console.error(`❌ Failed to retrieve important memories from Redis: ${error}`);
        // Fall back to in-memory retrieval
        return this.retrieveFromMemory(agent_id, 'importance', limit);
      }
    } else {
      // Get from in-memory cache
      return this.retrieveFromMemory(agent_id, 'importance', limit);
    }
  }

  /**
   * Retrieve memories from the in-memory cache
   */
  private retrieveFromMemory(
    agent_id: string,
    sort_by: 'timestamp' | 'importance',
    limit: number
  ): Memory[] {
    if (!this.inMemoryCache[agent_id]) {
      return [];
    }
    
    // Get all memories for the agent
    const agent_memories = Object.values(this.inMemoryCache[agent_id]);
    
    // Sort based on the specified field
    if (sort_by === 'timestamp') {
      agent_memories.sort((a, b) => b.created_at - a.created_at);
    } else if (sort_by === 'importance') {
      agent_memories.sort((a, b) => b.importance - a.importance);
    }
    
    // Return limited number of memories
    return agent_memories.slice(0, limit);
  }

  /**
   * Search memories based on content similarity (simple keyword matching)
   */
  public async searchMemories(
    agent_id: string,
    query: string,
    limit: number = 5
  ): Promise<Memory[]> {
    // Simple keyword search for now
    // Convert query to lowercase for case-insensitive matching
    const query_lower = query.toLowerCase();
    
    // Get all memories for the agent
    let memories: Memory[] = [];
    
    if (this.isRedisConnected && this.redisClient) {
      try {
        // Get all memory IDs for the agent
        const memory_ids = await this.redisClient.zRange(
          `memory_index:${agent_id}`, 
          0, 
          -1
        );
        
        if (!memory_ids || memory_ids.length === 0) {
          return [];
        }
        
        // Get the actual memories
        for (const memory_id of memory_ids) {
          const memory_json = await this.redisClient.get(`memory:${agent_id}:${memory_id}`);
          
          if (memory_json) {
            memories.push(JSON.parse(memory_json));
          }
        }
      } catch (error) {
        console.error(`❌ Failed to search memories in Redis: ${error}`);
        // Fall back to in-memory search
        return this.searchInMemory(agent_id, query_lower, limit);
      }
    } else {
      // Get from in-memory cache
      if (!this.inMemoryCache[agent_id]) {
        return [];
      }
      
      memories = Object.values(this.inMemoryCache[agent_id]);
    }
    
    // Filter memories that contain the query in the content
    const results = memories.filter(memory => 
      memory.content.toLowerCase().includes(query_lower)
    );
    
    // Sort by relevance (simple implementation)
    results.sort((a, b) => {
      // Prioritize exact matches, then importance
      const aExactMatch = a.content.toLowerCase().includes(query_lower);
      const bExactMatch = b.content.toLowerCase().includes(query_lower);
      
      if (aExactMatch && !bExactMatch) return -1;
      if (!aExactMatch && bExactMatch) return 1;
      
      // If both match or don't match, sort by importance
      return b.importance - a.importance;
    });
    
    // Return limited number of results
    return results.slice(0, limit);
  }

  /**
   * Search memories in the in-memory cache
   */
  private searchInMemory(
    agent_id: string,
    query: string,
    limit: number
  ): Memory[] {
    if (!this.inMemoryCache[agent_id]) {
      return [];
    }
    
    // Get all memories for the agent
    const agent_memories = Object.values(this.inMemoryCache[agent_id]);
    
    // Filter memories that contain the query in the content
    const results = agent_memories.filter(memory => 
      memory.content.toLowerCase().includes(query)
    );
    
    // Sort by relevance (simple implementation)
    results.sort((a, b) => {
      // Prioritize exact matches, then importance
      const aExactMatch = a.content.toLowerCase().includes(query);
      const bExactMatch = b.content.toLowerCase().includes(query);
      
      if (aExactMatch && !bExactMatch) return -1;
      if (!aExactMatch && bExactMatch) return 1;
      
      // If both match or don't match, sort by importance
      return b.importance - a.importance;
    });
    
    // Return limited number of results
    return results.slice(0, limit);
  }

  /**
   * Delete a memory
   */
  public async deleteMemory(agent_id: string, memory_id: string): Promise<boolean> {
    // Try to delete from Redis if connected
    if (this.isRedisConnected && this.redisClient) {
      try {
        // Remove from memory storage
        await this.redisClient.del(`memory:${agent_id}:${memory_id}`);
        
        // Remove from indices
        await this.redisClient.zRem(`memory_index:${agent_id}`, memory_id);
        await this.redisClient.zRem(`memory_importance:${agent_id}`, memory_id);
        
        console.log(`✅ Memory ${memory_id} deleted from Redis for agent ${agent_id}`);
        
        // Also remove from in-memory cache if it exists there
        this.deleteFromMemory(agent_id, memory_id);
        
        return true;
      } catch (error) {
        console.error(`❌ Failed to delete memory from Redis: ${error}`);
        // Fall back to in-memory deletion
        return this.deleteFromMemory(agent_id, memory_id);
      }
    } else {
      // Delete from in-memory cache
      return this.deleteFromMemory(agent_id, memory_id);
    }
  }

  /**
   * Delete memory from the in-memory cache
   */
  private deleteFromMemory(agent_id: string, memory_id: string): boolean {
    if (!this.inMemoryCache[agent_id]) {
      return false;
    }
    
    if (this.inMemoryCache[agent_id][memory_id]) {
      delete this.inMemoryCache[agent_id][memory_id];
      console.log(`✅ Memory ${memory_id} deleted from in-memory cache for agent ${agent_id}`);
      return true;
    }
    
    return false;
  }

  /**
   * Clear all memories for an agent
   */
  public async clearAgentMemories(agent_id: string): Promise<boolean> {
    // Try to clear from Redis if connected
    if (this.isRedisConnected && this.redisClient) {
      try {
        // Get all memory IDs for the agent
        const memory_ids = await this.redisClient.zRange(
          `memory_index:${agent_id}`, 
          0, 
          -1
        );
        
        // Delete all memories
        for (const memory_id of memory_ids) {
          await this.redisClient.del(`memory:${agent_id}:${memory_id}`);
        }
        
        // Delete indices
        await this.redisClient.del(`memory_index:${agent_id}`);
        await this.redisClient.del(`memory_importance:${agent_id}`);
        
        console.log(`✅ All memories cleared for agent ${agent_id}`);
        
        // Also clear from in-memory cache
        this.clearMemoryCache(agent_id);
        
        return true;
      } catch (error) {
        console.error(`❌ Failed to clear memories from Redis: ${error}`);
        // Fall back to in-memory clearing
        return this.clearMemoryCache(agent_id);
      }
    } else {
      // Clear from in-memory cache
      return this.clearMemoryCache(agent_id);
    }
  }

  /**
   * Clear agent memories from the in-memory cache
   */
  private clearMemoryCache(agent_id: string): boolean {
    if (this.inMemoryCache[agent_id]) {
      delete this.inMemoryCache[agent_id];
      console.log(`✅ All memories cleared from in-memory cache for agent ${agent_id}`);
      return true;
    }
    
    return false;
  }

  /**
   * Get a specific memory by ID
   */
  public async getMemory(agent_id: string, memory_id: string): Promise<Memory | null> {
    // Try to get from Redis if connected
    if (this.isRedisConnected && this.redisClient) {
      try {
        const memory_json = await this.redisClient.get(`memory:${agent_id}:${memory_id}`);
        
        if (memory_json) {
          return JSON.parse(memory_json);
        }
        
        console.warn(`⚠️ Memory ${memory_id} not found in Redis for agent ${agent_id}`);
      } catch (error) {
        console.error(`❌ Failed to get memory from Redis: ${error}`);
      }
    }
    
    // Fall back to in-memory cache
    if (
      this.inMemoryCache[agent_id] &&
      this.inMemoryCache[agent_id][memory_id]
    ) {
      return this.inMemoryCache[agent_id][memory_id];
    }
    
    return null;
  }

  /**
   * Update the importance score of a memory
   */
  public async updateMemoryImportance(
    agent_id: string,
    memory_id: string,
    importance: number
  ): Promise<boolean> {
    // Try to update in Redis if connected
    if (this.isRedisConnected && this.redisClient) {
      try {
        // Get the memory
        const memory_json = await this.redisClient.get(`memory:${agent_id}:${memory_id}`);
        
        if (!memory_json) {
          console.warn(`⚠️ Memory ${memory_id} not found in Redis for agent ${agent_id}`);
          return false;
        }
        
        const memory = JSON.parse(memory_json);
        memory.importance = importance;
        
        // Update the memory
        await this.redisClient.set(`memory:${agent_id}:${memory_id}`, JSON.stringify(memory));
        
        // Update the importance index
        await this.redisClient.zAdd(`memory_importance:${agent_id}`, {
          score: importance,
          value: memory_id
        });
        
        console.log(`✅ Importance updated to ${importance} for memory ${memory_id}`);
        
        // Also update in in-memory cache if it exists there
        if (
          this.inMemoryCache[agent_id] &&
          this.inMemoryCache[agent_id][memory_id]
        ) {
          this.inMemoryCache[agent_id][memory_id].importance = importance;
        }
        
        return true;
      } catch (error) {
        console.error(`❌ Failed to update memory importance in Redis: ${error}`);
        // Fall back to in-memory update
        return this.updateImportanceInMemory(agent_id, memory_id, importance);
      }
    } else {
      // Update in in-memory cache
      return this.updateImportanceInMemory(agent_id, memory_id, importance);
    }
  }

  /**
   * Update memory importance in the in-memory cache
   */
  private updateImportanceInMemory(
    agent_id: string,
    memory_id: string,
    importance: number
  ): boolean {
    if (
      !this.inMemoryCache[agent_id] ||
      !this.inMemoryCache[agent_id][memory_id]
    ) {
      return false;
    }
    
    this.inMemoryCache[agent_id][memory_id].importance = importance;
    console.log(`✅ Importance updated to ${importance} for memory ${memory_id} in in-memory cache`);
    return true;
  }

  /**
   * Close connections to external services
   */
  public async close(): Promise<void> {
    if (this.isRedisConnected && this.redisClient) {
      await this.redisClient.quit();
      console.log('✅ Redis client closed');
    }
  }
}

// Create singleton instance
const memoryService = new MemoryService();

export default memoryService;